cmd/agent/app/httpserver/server.go

```diff
@@ -0,0 +1,184 @@
+	"encoding/json"
+	"errors"
+	"fmt"
+	"net/http"
+	"strings"
+
+	"github.com/uber/jaeger-lib/metrics"
+
+	tSampling "github.com/uber/jaeger/thrift-gen/sampling"
+)
+
+const mimeTypeApplicationJSON = "application/json"
+
+var (
+	errBadRequest = errors.New("Bad request")
+)
+
+// NewHTTPServer creates a new server that hosts an HTTP/JSON endpoint for clients
+// to query for sampling strategies and baggage restrictions.
+func NewHTTPServer(hostPort string, manager ClientConfigManager, mFactory metrics.Factory) *http.Server {
+	handler := newHTTPHandler(manager, mFactory)
+	mux := http.NewServeMux()
+	mux.HandleFunc("/", func(w http.ResponseWriter, r *http.Request) {
+		handler.serveSamplingHTTP(w, r, true /* thriftEnums092 */)
+	})
+	mux.HandleFunc("/sampling", func(w http.ResponseWriter, r *http.Request) {
+		handler.serveSamplingHTTP(w, r, false /* thriftEnums092 */)
+	})
+	mux.HandleFunc("/baggage", func(w http.ResponseWriter, r *http.Request) {
+		handler.serveBaggageHTTP(w, r)
+	})
+	return &http.Server{Addr: hostPort, Handler: mux}
+}
+
+func newHTTPHandler(manager ClientConfigManager, mFactory metrics.Factory) *httpHandler {
+	handler := &httpHandler{manager: manager}
+	metrics.Init(&handler.metrics, mFactory, nil)
+	return handler
+}
+
+type httpHandler struct {
+	manager ClientConfigManager
+	metrics struct {
+		// Number of good sampling requests
+		SamplingRequestSuccess metrics.Counter `metric:"http-server.requests" tags:"result=ok,type=sampling"`
+
+		// Number of good sampling requests against the old endpoint / using Thrift 0.9.2 enum codes
+		LegacySamplingRequestSuccess metrics.Counter `metric:"http-server.requests" tags:"result=ok,type=sampling-legacy"`
+
+		// Number of good baggage requests
+		BaggageRequestSuccess metrics.Counter `metric:"http-server.requests" tags:"result=ok,type=baggage"`
+
+		// Number of bad requests (400s)
+		BadRequest metrics.Counter `metric:"http-server.requests" tags:"result=err,status=4xx"`
+
+		// Number of collector proxy failures
+		TCollectorProxyFailures metrics.Counter `metric:"http-server.requests" tags:"result=err,status=5xx,type=tcollector-proxy"`
+
+		// Number of bad responses due to malformed thrift
+		BadThriftFailures metrics.Counter `metric:"http-server.requests" tags:"result=err,status=5xx,type=thrift"`
+
+		// Number of failed response writes from http server
+		WriteFailures metrics.Counter `metric:"http-server.requests" tags:"result=err,status=5xx,type=write"`
+	}
+}
+
+func (h *httpHandler) serviceFromRequest(w http.ResponseWriter, r *http.Request) (string, error) {
+	services := r.URL.Query()["service"]
+	if len(services) != 1 {
+		h.metrics.BadRequest.Inc(1)
+		http.Error(w, "'service' parameter must be provided once", http.StatusBadRequest)
+		return "", errBadRequest
+	}
+	return services[0], nil
+}
+
+func (h *httpHandler) writeJSON(w http.ResponseWriter, json []byte) error {
+	w.Header().Add("Content-Type", mimeTypeApplicationJSON)
+	if _, err := w.Write(json); err != nil {
```

Comment:
it only errors if the http server fails to accept the write which we have tests for